Update to accommodate MvvmCross 6

sfmskywalker opened the pull request:
This PR updates the referenced MvvmCross packages to the single MvvmCross 6.0.1 package, and adds support for creating child containers.

Because `AutofacMvxIocProvider` inherits from `MvxSingleton`, I had to create a separate class that provides the actual implementation in order to be able to create child IoC providers.

I also added a unit test ensuring that child containers can be created and behave as expected.

The project now supports.NET standard 2.0, and the unit test project has been converted to .NET Core 2.0.

Fixes #11

SipkeSchoorstraRTL committed:
Updated old project format.

SipkeSchoorstraRTL committed:
Fixed style cop complaints.

Commit:
Removed old files.

Commit:
Updated Autofac.Build.psm1 to use .csproj instead of the old .xproj.

Commit:
Updated Autofac.Build.psm1 dotnet-cli download path.

Commit:
Disabled warning CS0618 for the time being.

sfmskywalker commented:
I might need some help fixing the appveyor build - not sure how that's all setup and why it fails.

Commit:
Updated package references.

alexmg commented on src/Autofac.Extras.MvvmCross/Autofac.Extras.MvvmCross.csproj:
```diff
@@ -0,0 +1,62 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Autofac" Version="4.8.1" />
```
Is there a lower version that can be targeted?

sfmskywalker replied:
Sure, I'll set it back to 4.0.0-* as it was before.

alexmg commented on src/Autofac.Extras.MvvmCross/Autofac.Extras.MvvmCross.csproj:
```diff
@@ -0,0 +1,62 @@
+
+  <PropertyGroup>
+    <Description>Autofac extension supporting MvvmCross dependency injection.</Description>
+    <VersionPrefix>4.0.2</VersionPrefix>
```
There are breaking changes here so might need to go for `5.0.0`.

sfmskywalker replied:
Good point, I'll update this to `5.0.0`.

alexmg commented on src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:
```diff
@@ -146,23 +123,7 @@ public virtual void CallbackWhenRegistered<T>(Action action)
-                    action();
-                }
-            };
+            this.provider.CallbackWhenRegistered(type, action);
```
Can you confirm the argument checks are not required?

sfmskywalker replied:
Confirmed - I'm simply delegating the calls to the new `ChildAutofacMvxIocProvider` class that provides the actual implementation. Ideally we would need only one class that implements `IMvxIoCProvider`, but the "root" provider needs to be registered as a singleton with MvvmCross. Perhaps there's another way to do that other than deriving from `MvxSingleton<T>` That would be much better actually. I'll see if this is possible or not.

sfmskywalker replied:
Not possible unfortunately. We need a class that derives from `MvxSingleton<TInterface>`, which sets a static `Instance` property (with a private setter) and adds itself to a private static list of `Singletons`.  Furthermore, MvvmCross access `MvxSingleton<IMvxIoCProvider>` from many places.

alexmg commented on src/Autofac.Extras.MvvmCross/AutofacMvxIocProvider.cs:
```diff
@@ -284,25 +240,7 @@ public virtual T GetSingleton<T>()
-            if (registration.Sharing != InstanceSharing.Shared || !(registration.Lifetime is RootScopeLifetime))
-            {
-                // Ensure the dependency is registered as a singleton WITHOUT resolving the dependency twice.
-                throw new DependencyResolutionException(string.Format(CultureInfo.CurrentCulture, AutofacMvxIocProviderResources.TypeNotRegisteredAsSingleton, type));
```
Please delete resource if no longer used.

sfmskywalker replied:
It's still used - in `ChildAutofacMvxIocProvider`.

alexmg commented on src/Autofac.Extras.MvvmCross/AutofacPropertyInjectorOptions.cs:
```diff
@@ -1,8 +1,12 @@
-﻿using Autofac.Core;
-using MvvmCross.Platform.IoC;
+﻿// <copyright file="AutofacPropertyInjectorOptions.cs" company="PlaceholderCompany">
```
Please use the regular copyright header.

sfmskywalker replied:
I was looking at the develop branch to try and figure out what the regular copyright header is supposed to be, but I didn't see it anywhere. Should I perhaps just remove this copyright placeholder?

sfmskywalker replied:
When I remove the header, StyleCop complaints about it. Since I don't get this error with other projects, I'm assuming StyleCop is enabled for this project, is that correct?

Here's what I'm seeing when I remove the copyright header:

![image](https://user-images.githubusercontent.com/938393/40889636-d807613a-676a-11e8-92bf-a0bfd90f28d8.png)

alexmg commented on src/Autofac.Extras.MvvmCross/ChildAutofacMvxIocProvider.cs:
```diff
@@ -0,0 +1,740 @@
+﻿// <copyright file="ChildAutofacMvxIocProvider.cs" company="PlaceholderCompany">
```
Please use the regular copyright header.

alexmg commented on src/Autofac.Extras.MvvmCross/IAutofacPropertyInjectorOptions.cs:
```diff
@@ -1,8 +1,12 @@
-﻿using Autofac.Core;
-using MvvmCross.Platform.IoC;
+﻿// <copyright file="IAutofacPropertyInjectorOptions.cs" company="PlaceholderCompany">
```
Please use the regular copyright header.

alexmg commented on src/Autofac.Extras.MvvmCross/Properties/AssemblyInfo.cs:
```diff
@@ -1,3 +1,7 @@
+// <copyright file="AssemblyInfo.cs" company="PlaceholderCompany">
```
Please use the regular copyright header.

alexmg reviewed (changes requested):
Thanks for the PR @sfmskywalker. It's a decent size change but from what I can see it looks alright apart from a few cosmetic things. Let's get it cleaned up and figure out what the build issue is. 👍 This will need to be a `5.0.0` release of the package too.

sfmskywalker commented:
Thanks for the help and review @alexmg !

Commit:
PR feedback.

Commit:
Updating to support breaking changes in MvvmCross Beta 1

Commit:
Added auto-generated code file headers to appease StyleCop.

Commit:
Updating package reference

DaRosenberg committed:
Updated Autofac package reference.

DaRosenberg committed:
Updated to MvvmCross 6.2.0-beta.2.

DaRosenberg committed:
Cleaned up conflicting references in project file.

michaldudak committed:
Changed the .NET Core installation channel

Added the -Channel Current switch to .NET Core installation script to install the latest version instead of LTS (currently 1.1).

michaldudak commented:
Hi there, I've created a pull request to this pull request's branch with a change that may fix the AppVeyor's build (https://github.com/IDeliverable/Autofac.Extras.MvvmCross/pull/2). This will install .NET Core 2.1 instead of 1.1. This allows the build to run.

michaldudak committed:
Fixed the test project's compilation errors

Commit:
Merge pull request #2 from michaldudak/patch-1

Changed the .NET Core installation channel

sfmskywalker commented:
Many thanks!

michaldudak commented on src/Autofac.Extras.MvvmCross/Autofac.Extras.MvvmCross.csproj:
```diff
@@ -0,0 +1,47 @@
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+    <PackageReference Include="Autofac" Version="4.8.1" />
+    <PackageReference Include="MvvmCross" Version="6.2.0-beta.2" />
```
Why does this project depend on a prerelease version? Wouldn't it be better to aim for the lowest possible version or at least a stable one?

sfmskywalker replied:
The reason it's depending on MvvmCross 6.2.0-beta.2 is because my project depends on features provided by the upcoming MvvmCross 6.2.0 release. I figured since I'm updating our clone of `Autofac.Extras.MvvmCross` anyway, might as well share it via a PR so others might benefit from it. 

But I understand if this won't be released until MvvmCross 6.2.0 is also released, at which point we can update the package reference here.

miszu replied:
hey there, thanks a lot for this code. So 6.2 is already released, can I help anyhow? we use this package and would love to have a stable version referenced in our project.

tillig replied:
It looks like we should be able to pull this in and make a few tweaks (version numbers, etc.) to get it out the door.

(If folks _really_ want to help, [we're still looking for project owners for extensions like this](https://www.paraesthesia.com/archive/2018/04/02/autofac-looking-for-extension-owners/)...)

tillig commented:
This is built and ready for test as version `5.0.0-develop-00266` on the Autofac MyGet feed: `https://www.myget.org/F/autofac/api/v3/index.json`

Please give it a shot and verify it works. I'm not an MvvmCross user so as long as the unit tests pass, I have to call it good.

If I don't hear anything by the end of the week (Friday, Oct. 5) I'll just push it to NuGet and folks can file issues if they find bugs.

miszu commented:
I tried the `5.0.0-develop-00266` and it just works with MVVM Cross 6.2, no changes needed at all. From my perspective, it's ready to be published, but of course there may be some features we don't use and because of that not tested. Thanks a lot guys!

tillig commented:
Version 5.0.0 has been released.
